rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,615 @@
+  void (*wait_for_event)(
+      SP_Device* executor, SP_Stream stream,
+      SP_Event event, TF_Status* status);
+
+  /*** TIMER CALLBACKS ***/
+  // Creates TF_Timer. Allocates timer resources on the underlying platform and initializes its
+  // internals, setting `timer` output variable. Sets values in `timer_fns` struct.
+  void (*create_timer)(SP_Device* executor, SP_Timer* timer, SP_TimerFns* timer_fns, TF_Status* status);
+
+  // Destroy timer and deallocates timer resources on the underlying platform.
+  void (*destroy_timer)(SP_Device* executor, SP_Timer timer, SP_TimerFns* timer_fns);
+
+  // Records a start event for an interval timer.
+  TF_BOOL (*start_timer)(
+      SP_Device* executor, SP_Stream stream, SP_Timer timer);
+
+
+  // Records a stop event for an interval timer.
+  TF_BOOL (*stop_timer)(
+      SP_Device* executor, SP_Stream stream, SP_Timer timer);
+
+  /*** MEMCPY CALLBACKS ***/
+  // Enqueues a memcpy operation onto stream, with a host destination location
+  // host_dst and a device memory source, with target size size.
+  TF_BOOL (*memcpy_dtoh)(
+      SP_Device* executor, SP_Stream stream,
+      void* host_dst,
+      const SE_DeviceMemoryBase* device_src,
+      uint64_t size);
+
+  // Enqueues a memcpy operation onto stream, with a device destination location
+  // and a host memory source, with target size size
+  TF_BOOL (*memcpy_htod)(
+      SP_Device* executor, SP_Stream stream,
+      SE_DeviceMemoryBase* device_dst,
+      const void* host_src, uint64_t size);
+      
+  // Enqueues a memcpy operation onto stream, with a device destination
+  // location and a device memory source, with target size `size`.
+  void (*memcpy_dtod)(const SP_Device* device, SP_Stream stream,
+                      SP_DeviceMemoryBase* device_dst,
+                      const SP_DeviceMemoryBase* device_src, uint64_t size,
+                      TF_Status* status);
+      
+  // Blocks the caller while a data segment of the given size is
+  // copied from the device source to the host destination.
+  TF_BOOL (*sync_memcpy_dtoh)(
+      SP_Device* executor,
+      void* host_dst,
+      const SE_DeviceMemoryBase* device_src,
+      uint64_t size);
+
+  // Blocks the caller while a data segment of the given size is
+  // copied from the host source to the device destination.
+  TF_BOOL (*sync_memcpy_htod)(
+      SP_Device* executor,
+      SE_DeviceMemoryBase* device_dst,
+      const void* host_src, uint64_t size);
+      
+  // Blocks the caller while a data segment of the given size is copied from the
+  // device source to the device destination.
+  void (*sync_memcpy_dtod)(const SP_Device* device,
+                           SP_DeviceMemoryBase* device_dst,
+                           const SP_DeviceMemoryBase* device_src, uint64_t size,
+                           TF_Status* status);
+
+  // Causes the host code to synchronously wait for the event to complete.
+  void (*block_host_for_event)(
```

Comment:
Hi, @annarev  It seems that block_host_until_done is removed, can we keep this API? block_host_until_done is a  more large-grained API, it provides more flexibility for the backend to implement, and lots of device runtime already provide the capability to match this API(CUDA: cuStreamSynchronize(stream),SYCL queue.wait() ). I saw current implementation is composed of create_event+ record_event + block_host_for_event + delete event, that means each back-end needs to implements all of these API, which limits the flexibility, and I am not sure whether it will bring some performance penalty (for DPC++, we need to enqueue a  barrier  to retrieve the event)

Reply:
Added here now:
https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/experimental/stream_executor/stream_executor.h;l=331?q=stream_executor.h